frontend/src/views/instructor/ddahs/import.tsx

```diff
@@ -0,0 +1,319 @@
+import React from "react";
+import { Alert } from "react-bootstrap";
+import { FaArrowRight } from "react-icons/fa";
+import { useSelector } from "react-redux";
+import { assignmentsSelector, applicantsSelector } from "../../../api/actions";
+import { ddahsSelector, upsertDdahs } from "../../../api/actions/ddahs";
+import {
+    Ddah,
+    Assignment,
+    Applicant,
+    MinimalDdah,
+    RawDuty,
+} from "../../../api/defs/types";
+import { ImportActionButton } from "../../../components/import-button";
+import { DiffSpec, diffImport, getChanged } from "../../../libs/diffs";
+import { normalizeDdahImports } from "../../../libs/import-export";
+import { useThunkDispatch } from "../../../libs/thunk-dispatch";
+
+enum DiffType {
+    Unchanged = "UNCHANGED",
+    Created = "CREATED",
+    Updated = "UPDATED",
+    Deleted = "DELETED",
+}
+
+type MinimalDuty = {
+    description: string;
+    hours: number;
+};
+
+interface DdahUpdate {
+    updatedDdah: Ddah;
+    oldDdah: Ddah;
+}
+
+interface DutyDiff {
+    oldDuty?: MinimalDuty;
+    newDuty?: MinimalDuty;
+    status: DiffType;
+}
+
+function getDutiesDiff(oldDuties: RawDuty[], newDuties: RawDuty[]) {
+    let changes: DutyDiff[] = [];
+    let dutiesChanged = false;
+
+    oldDuties.forEach((oldDuty) => {
+        let oldDutyDeleted = true;
+
+        newDuties.forEach((newDuty, index) => {
+            if (
+                oldDuty.description === newDuty.description &&
+                oldDuty.hours === newDuty.hours
+            ) {
+                changes.push({ oldDuty, newDuty, status: DiffType.Unchanged });
+                newDuties.splice(index, 1);
+                oldDutyDeleted = false;
+            } else if (
+                oldDuty.description === newDuty.description &&
+                oldDuty.hours !== newDuty.hours
+            ) {
+                changes.push({ oldDuty, newDuty, status: DiffType.Updated });
+                newDuties.splice(index, 1);
+                oldDutyDeleted = false;
+                dutiesChanged = true;
+            }
+        });
+
+        // If we did not find old duties to match the new one - mark it as created
+        if (oldDutyDeleted) {
+            changes.push({ oldDuty, status: DiffType.Deleted });
+            dutiesChanged = true;
+        }
+    });
+
+    // The new duties that are left were just created - add them
+    newDuties.forEach((duty) =>
+        changes.push({ newDuty: duty, status: DiffType.Created })
+    );
+
+    // Do not return changes if everything is unchanged
+    return dutiesChanged ? changes : [];
+}
+
+function dutiesDiffList(changes: DutyDiff[]) {
+    return (
+        <ul>
+            {changes.map(({ oldDuty, newDuty, status }) => {
+                switch (status) {
```

Comment:
This makes me unhappy...Make a `DutyItem` component which contains this logic instead. I don't like seeing switch statement logic nested deep inside the JSX.

Reply:
If I take out the `switch` here, the whole component will contain of a `<ul>` followed by a `map`. The `switch` here is one level nested, so I was thinking it is fine? Let me know what you think :slightly_smiling_face:

Reply:
Nope. Change it.